README.md

# LBFGS

[![Build Status](https://travis-ci.org/ybyygu/rust-lbfgs.svg?branch=master)](https://travis-ci.org/ybyygu/rust-lbfgs)
[![GPL3 licensed](https://img.shields.io/badge/license-GPL3-blue.svg)](./LICENSE)

Fast and safe Rust implementation of LBFGS and OWL-QN algorithms ported from
Naoaki Okazaki's C library [libLBFGS](http://chokkan.org/software/liblbfgs/).

Check [rust-liblbfgs](https://github.com/ybyygu/rust-liblbfgs) for a working wrapper around the original C codes.


# Motivation

-   Bring native LBFGS implementation to Rust community.
-   Learn how a great optimization algorithm is implemented in real world.
-   Learn how to "replace the jet engine while still flying" [URL](http://jensimmons.com/post/jan-4-2017/replacing-jet-engine-while-still-flying)
-   Make it more maintainable with Rust high level abstraction.
-   Improve it to meet my needs for computational chemistry.


# Todo

-   [ ] Parallel with rayon
-   [ ] SIMD support
-   [X] add option to disable line search for gradient only optimization
-   [X] Fix issues inherited from liblbfgs [URL](https://github.com/chokkan/liblbfgs/pulls)


# Features

-   Clean and safe Rust implementation.
-   OWL-QN algorithm.
-   Closure based callback interfaces.
-   Damped L-BFGS algorithm.


# Usage

    // 0. Import the lib
    use liblbfgs::lbfgs;
    
    const N: usize = 100;
    
    // 1. Initialize data
    let mut x = [0.0 as f64; N];
    for i in (0..N).step_by(2) {
        x[i] = -1.2;
        x[i + 1] = 1.0;
    }
    
    // 2. Defining how to evaluate function and gradient
    let evaluate = |x: &[f64], gx: &mut [f64]| {
        let n = x.len();
    
        let mut fx = 0.0;
        for i in (0..n).step_by(2) {
            let t1 = 1.0 - x[i];
            let t2 = 10.0 * (x[i + 1] - x[i] * x[i]);
            gx[i + 1] = 20.0 * t2;
            gx[i] = -2.0 * (x[i] * gx[i + 1] + t1);
            fx += t1 * t1 + t2 * t2;
        }
    
        Ok(fx)
    };
    
    let prb = lbfgs()
        .with_max_iterations(5)
        .with_orthantwise(1.0, 0, 99) // enable OWL-QN
        .minimize(
            &mut x,                   // input variables
            evaluate,                 // define how to evaluate function
            |prgr| {                  // define progress monitor
                println!("iter: {:}", prgr.niter);
                false                 // returning true will cancel optimization
            }
        )
        .expect("lbfgs owlqn minimize");
    
    println!("fx = {:}", prb.fx);

The callback functions are native Rust FnMut closures, possible to
capture/change variables in the environment.

Full codes with comments are available in examples/sample.rs.

Run the example:

    cargo run --example sample



CHANGELOG.md

# v0.0.15

-   <span class="timestamp-wrapper"><span class="timestamp">[2020-02-10 Mon] </span></span> remove quicli dependency
-   <span class="timestamp-wrapper"><span class="timestamp">[2020-02-10 Mon] </span></span> rename package as liblbfgs


# v0.0.14

-   <span class="timestamp-wrapper"><span class="timestamp">[2019-03-22 Fri] </span></span> Damped LBFGS algorithm
-   <span class="timestamp-wrapper"><span class="timestamp">[2019-03-22 Fri] </span></span> Gradient only optimization.


# v0.0.13

-   <span class="timestamp-wrapper"><span class="timestamp">[2019-01-02 Wed] </span></span> new example: optimization of lj38 cluster
-   <span class="timestamp-wrapper"><span class="timestamp">[2018-12-21 Fri] </span></span> fixed progress report issue.


# v0.0.12

-   <span class="timestamp-wrapper"><span class="timestamp">[2018-12-18 Tue] </span></span> new parameter to control max allowed evaluations: max\_evaluations
-   <span class="timestamp-wrapper"><span class="timestamp">[2018-12-18 Tue] </span></span> force to set progress callback function in minimize method.


# v0.0.11

-   <span class="timestamp-wrapper"><span class="timestamp">[2018-12-17 Mon] </span></span> assign lbfgs parameters using builder pattern
-   <span class="timestamp-wrapper"><span class="timestamp">[2018-12-16 Sun] </span></span> fixed important issues inherited from liblibfgs (#2, #3)
-   <span class="timestamp-wrapper"><span class="timestamp">[2018-12-15 Sat] </span></span> remove wolfe parameter from LineSearch (use gtol instead)
-   <span class="timestamp-wrapper"><span class="timestamp">[2018-12-14 Fri] </span></span> LineSearchParm is renamed as LineSearch


# v0.0.10

-   <span class="timestamp-wrapper"><span class="timestamp">[2018-12-12 Wed] </span></span> add Orthantwise struct to represent all orthantwise parameters (see: param.owlqn)
-   <span class="timestamp-wrapper"><span class="timestamp">[2018-12-12 Wed] </span></span> use param.orthantwise option to enable/disable OWL-QN algorithm
-   <span class="timestamp-wrapper"><span class="timestamp">[2018-12-11 Tue] </span></span> new line mod


# v0.0.9

-   <span class="timestamp-wrapper"><span class="timestamp">[2018-12-11 Tue] </span></span> add math mod
-   <span class="timestamp-wrapper"><span class="timestamp">[2018-12-10 Mon] </span></span> remove all unsafe codes


# v0.0.8

-   <span class="timestamp-wrapper"><span class="timestamp">[2018-12-10 Mon] </span></span> update quicli to v0.4
-   <span class="timestamp-wrapper"><span class="timestamp">[2018-12-06 Thu] </span></span> add test for OWL-QN
-   <span class="timestamp-wrapper"><span class="timestamp">[2018-12-05 Wed] </span></span> clean up some unsafe codes


# v0.0.7

-   <span class="timestamp-wrapper"><span class="timestamp">[2018-11-26 Mon] </span></span> Rust codes translated from liblbfgs/c using c2rust


# v0.0.6

-   <span class="timestamp-wrapper"><span class="timestamp">[2018-11-11 Sun] </span></span> new construct method with epsilon
    
        let lbfgs = LBFGS::new(fmax);


# v0.0.5

-   <span class="timestamp-wrapper"><span class="timestamp">[2018-11-16 Fri] </span></span> callback interfaces to liblbfgs/c using closures



src/lib.rs
// docs

//
//  Copyright (c) 1990, Jorge Nocedal
//  Copyright (c) 2007-2010 Naoaki Okazaki
//  Copyright (c) 2018-2019 Wenping Guo
//  All rights reserved.
//
//! Limited memory BFGS (L-BFGS) algorithm ported from liblbfgs
//!
//! # Example
//! ```
//! // 0. Import the lib
//! use liblbfgs::lbfgs;
//! 
//! const N: usize = 100;
//! 
//! // 1. Initialize data
//! let mut x = [0.0 as f64; N];
//! for i in (0..N).step_by(2) {
//!     x[i] = -1.2;
//!     x[i + 1] = 1.0;
//! }
//! 
//! // 2. Defining how to evaluate function and gradient
//! let evaluate = |x: &[f64], gx: &mut [f64]| {
//!     let n = x.len();
//! 
//!     let mut fx = 0.0;
//!     for i in (0..n).step_by(2) {
//!         let t1 = 1.0 - x[i];
//!         let t2 = 10.0 * (x[i + 1] - x[i] * x[i]);
//!         gx[i + 1] = 20.0 * t2;
//!         gx[i] = -2.0 * (x[i] * gx[i + 1] + t1);
//!         fx += t1 * t1 + t2 * t2;
//!     }
//! 
//!     Ok(fx)
//! };
//! 
//! let prb = lbfgs()
//!     .with_max_iterations(5)
//!     .with_orthantwise(1.0, 0, 99) // enable OWL-QN
//!     .minimize(
//!         &mut x,                   // input variables
//!         evaluate,                 // define how to evaluate function
//!         |prgr| {                  // define progress monitor
//!             println!("iter: {:}", prgr.niter);
//!             false                 // returning true will cancel optimization
//!         }
//!     )
//!     .expect("lbfgs owlqn minimize");
//! 
//! println!("fx = {:}", prb.fx);
//! ```

// base

// [[file:~/Workspace/Programming/gosh-rs/lbfgs/lbfgs.note::*base][base:1]]
use crate::core::*;

mod lbfgs;
pub mod line;
pub mod math;
pub use crate::lbfgs::*;

pub(crate) mod core {
    pub use anyhow::*;
    pub use log::{debug, error, info, trace, warn};
}
// base:1 ends here

// lbfgs

// [[file:~/Workspace/Programming/gosh-rs/lbfgs/lbfgs.note::*lbfgs][lbfgs:1]]
pub fn lbfgs<F>() -> LBFGS<F>
where
    F: FnMut(&[f64], &mut [f64]) -> Result<f64>,
{
    LBFGS::default()
}
// lbfgs:1 ends here

// tests
// test functions

// [[file:~/Workspace/Programming/gosh-rs/lbfgs/lbfgs.note::*tests][tests:1]]
/// Default test function (rosenbrock) adopted from liblbfgs sample.c
pub fn default_evaluate() -> impl FnMut(&[f64], &mut [f64]) -> Result<f64> {
    move |arr_x: &[f64], gx: &mut [f64]| {
        let n = arr_x.len();

        let mut fx = 0.0;
        for i in (0..n).step_by(2) {
            let t1 = 1.0 - arr_x[i];
            let t2 = 10.0 * (arr_x[i + 1] - arr_x[i] * arr_x[i]);
            gx[i + 1] = 20.0 * t2;
            gx[i] = -2.0 * (arr_x[i] * gx[i + 1] + t1);
            fx += t1 * t1 + t2 * t2;
        }

        Ok(fx)
    }
}

/// Default progress monitor adopted from liblbfgs sample.c
///
/// # Notes
///
/// * Returning true will cancel the optimization process.
///
pub fn default_progress() -> impl FnMut(&Progress) -> bool {
    move |prgr| {
        println!("Iteration {}, Evaluation {}:", prgr.niter, prgr.neval);
        println!(
            " fx = {:-12.6} xnorm = {:-12.6}, gnorm = {:-12.6}, ls = {}, step = {}",
            prgr.fx, prgr.xnorm, prgr.gnorm, prgr.ncall, prgr.step
        );

        false
    }
}
// tests:1 ends here
